### PR TITLE
Support string type in segcore

### PR DESCRIPTION
**Pull request by xige-16**
fix issue: #15888 
/kind feature

### CODE REVIEW

File: internal/core/src/common/FieldMeta.h [hunk 1]
```diff
@@ -128,6 +139,11 @@ class FieldMeta {
         Assert(!is_vector());
     }
 
+    FieldMeta(const FieldName& name, FieldId id, DataType type, int64_t max_length_per_row)
+        : name_(name), id_(id), type_(type), string_info_(StringInfo{max_length_per_row}) {
```
Comment (cydrain) [inline internal/core/src/common/FieldMeta.h:143]: string_info_(max_length_per_row) -- can write in this way ?

File: internal/core/src/common/Schema.h [hunk 2]
```diff
@@ -152,29 +142,33 @@ class Schema {
...
-        AssertInfo(!id_offsets_.count(field_meta.get_id()), "duplicated field id");
-        id_offsets_.emplace(field_meta.get_id(), offset);
+        auto field_name = field_meta.get_name();
+        auto field_id = field_meta.get_id();
+        AssertInfo(!name_ids_.count(field_name), "duplicated field name");
+        AssertInfo(!id_names_.count(field_id), "duplicated field id");
+        name_ids_.emplace(field_name, field_id);
+        id_names_.emplace(field_id, field_name);
+
+        fields_.emplace(field_id, std::move(field_meta));
```
Comment (Presburger) [inline internal/core/src/common/Schema.h:152]: 这边field_meta建议不要调std::move转右值，下面这个变量还继续使用了。

File: internal/core/src/common/Schema.cpp [hunk 3]
```diff
@@ -69,25 +68,26 @@ Schema::ParseFrom(const milvus::proto::schema::CollectionSchema& schema_proto) {
                 auto metric_type = GetMetricType(index_map.at("metric_type"));
                 schema->AddField(name, field_id, data_type, dim, metric_type);
             }
+        } else if (datatype_is_string(data_type)) {
+            auto type_map = RepeatedKeyValToMap(child.type_params());
+            AssertInfo(type_map.count("max_length_per_row"), "max_length_per_row not found");
+            auto max_len = boost::lexical_cast<int64_t>(type_map.at("max_length_per_row"));
```
Comment (cydrain) [inline internal/core/src/common/Schema.cpp:74]: we'd better define a const char* instead of using "max_length_per_row" every where

File: internal/core/src/common/Schema.h [hunk 4]
```diff
@@ -100,50 +97,43 @@ class Schema {
...
-        Assert(id_offsets_.count(field_id));
-        return id_offsets_.at(field_id);
+    FieldId
+    get_field_id(const FieldName& field_name) const {
+        AssertInfo(name_ids_.count(field_name), "Cannot find field_name");
+        return name_ids_.at(field_name);
     }
 
-    const std::vector<FieldMeta>&
+    const std::unordered_map<FieldId, FieldMeta>&
```
Comment (cydrain) [inline internal/core/src/common/Schema.h:117]: it's better to move this change to another PR

File: internal/core/src/common/Types.cpp [hunk 5]
```diff
@@ -55,4 +55,24 @@ MetricTypeToName(MetricType metric_type) {
     return metric_bimap.right.at(metric_type);
 }
 
+bool
+IsPrimaryKeyDataType(DataType data_type) {
+    return data_type == engine::DataType::INT64 || data_type == DataType::VarChar;
```
Comment (cydrain) [inline internal/core/src/common/Types.cpp:60]: DataType::STRING cannot be primary key data type ?

File: internal/core/src/common/Types.cpp [hunk 6]
```diff
@@ -55,4 +55,24 @@ MetricTypeToName(MetricType metric_type) {
...
+}
+
+bool
+IsValidPrimaryKey(PkType pk) {
+    // pk type is int64
+    if (pk.index() == 1 && std::get<int64_t>(pk) != -1) {
```
Comment (cydrain) [inline internal/core/src/common/Types.cpp:66]: don't use "-1", use INVALID_ID

File: internal/core/src/common/Utils.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    for (int i = 0; i < prefix.length(); i++) {
+        if (prefix[i] != str[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+inline bool
+PostfixMatch(const std::string& str, const std::string& postfix) {
```
Comment (cydrain) [inline internal/core/src/common/Utils.h:31]: correct word is "Suffix"

File: internal/core/src/common/Utils.h [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+#include <string>
+
+namespace milvus {
+inline bool
+PrefixMatch(const std::string& str, const std::string& prefix) {
+    if (prefix.length() > str.length()) {
+        return false;
+    }
+    for (int i = 0; i < prefix.length(); i++) {
```
Comment (cydrain) [inline internal/core/src/common/Utils.h:22]: can use strncmp(str.cstr(), prefix.cstr(), str.length())

File: internal/core/src/common/Utils.h [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    }
+    return true;
+}
+
+inline bool
+PostfixMatch(const std::string& str, const std::string& postfix) {
+    if (postfix.length() > str.length()) {
+        return false;
+    }
+    int i = postfix.length() - 1;
```
Comment (cydrain) [inline internal/core/src/common/Utils.h:35]: can use strncmp()

File: internal/core/src/segcore/FieldIndexing.cpp [hunk 10]
```diff
@@ -142,6 +149,8 @@ CreateIndex(const FieldMeta& field_meta, const SegcoreConfig& segcore_config) {
             return std::make_unique<ScalarFieldIndexing<float>>(field_meta, segcore_config);
         case DataType::DOUBLE:
             return std::make_unique<ScalarFieldIndexing<double>>(field_meta, segcore_config);
+        case DataType::VarChar:
```
Comment (cydrain) [inline internal/core/src/segcore/FieldIndexing.cpp:152]: why not use "VARCHAR" just like other types ?

File: internal/core/src/utils/Types.h [hunk 11]
```diff
@@ -41,6 +41,7 @@ enum class DataType {
     DOUBLE = 11,
 
     STRING = 20,
+    VarChar = 21,
```
Comment (cydrain) [inline internal/core/src/utils/Types.h:44]: use VARCHAR

File: internal/core/thirdparty/knowhere/CMakeLists.txt [hunk 12]
```diff
@@ -27,7 +27,7 @@ macro(build_knowhere)
 
     set(KNOWHERE_CMAKE_ARGS
             ${EP_COMMON_CMAKE_ARGS}
-            -DCMAKE_BUILD_TYPE=Release
```
Comment (cydrain) [inline internal/core/thirdparty/knowhere/CMakeLists.txt:30]: change back

File: internal/core/unittest/bench/bench_indexbuilder.cpp [hunk 13]
```diff
@@ -59,7 +59,7 @@ IndexBuilder_build(benchmark::State& state) {
 
     auto is_binary = state.range(2);
     auto dataset = GenDataset(NB, metric_type, is_binary);
-    auto xb_data = dataset.get_col<float>(0);
+    auto xb_data = dataset.get_col<float>(milvus::FieldId(100));
```
Comment (cydrain) [inline internal/core/unittest/bench/bench_indexbuilder.cpp:62]: use FieldUserID instad of 100

File: internal/proxy/plan_parser.go [hunk 14]
```diff
@@ -237,6 +237,10 @@ func getCompareOpType(opStr string, reverse bool) (op planpb.OpType) {
 		op = planpb.OpType_Equal
 	case "!=":
 		op = planpb.OpType_NotEqual
+	case "startsWith":
```
Comment (cydrain) [inline internal/proxy/plan_parser.go:240]: can use '^' for "begin from", and '$' for "end with" ?

File: internal/core/src/common/FieldMeta.h [hunk 15]
```diff
@@ -139,13 +155,26 @@ class FieldMeta {
         return type_ == DataType::VECTOR_BINARY || type_ == DataType::VECTOR_FLOAT;
     }
 
+    bool
+    is_string() const {
```
Comment (cydrain) [inline internal/core/src/common/FieldMeta.h:159]: this API seems duplicate with datatype_is_string()

File: internal/core/src/index/StringIndexSort.h [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    BuildWithDataset(const DatasetPtr& dataset) override {
+        auto size = dataset->Get<int64_t>(knowhere::meta::ROWS);
+        auto data = dataset->Get<const void*>(knowhere::meta::TENSOR);
+        proto::schema::StringArray arr;
+        arr.ParseFromArray(data, size);
+
+        {
+            // TODO: optimize here. avoid memory copy.
+            std::vector<std::string> vecs{arr.data().begin(), arr.data().end()};
+            Build(arr.data().size(), vecs.data());
```
Comment (cydrain) [inline internal/core/src/index/StringIndexSort.h:34]: Generally "data" is the first parameter, "size" is behind

File: internal/core/src/index/StringIndexSort.h [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+        }
+
+        {
+            // TODO: test this way.
+            // auto strs = (const std::string*)arr.data().data();
+            // Build(arr.data().size(), strs);
+        }
+    }
+
+    const TargetBitmapPtr
```
Comment (cydrain) [inline internal/core/src/index/StringIndexSort.h:44]: it's strange to mark a return type as "const".
"const" should be marked where the return value is assigned, such as:
const TargetBitmapPtr res = Query(ds);

File: internal/core/src/index/StringIndexSort.h [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    Query(const DatasetPtr& dataset) override {
+        auto op = dataset->Get<OperatorType>(OPERATOR_TYPE);
+        if (op == PrefixMatchOp) {
+            auto prefix = dataset->Get<std::string>(PREFIX_VALUE);
+            return PrefixMatch(prefix);
+        }
+        return ScalarIndex<std::string>::Query(dataset);
+    }
+
+    const TargetBitmapPtr
```
Comment (cydrain) [inline internal/core/src/index/StringIndexSort.h:54]: same as above

File: internal/core/src/segcore/ConcurrentVector.h [hunk 19]
```diff
@@ -84,6 +90,113 @@ class VectorBase {
...
+            case DataType::INT8: {
+                auto src_data = data->scalars().int_data().data();
+                std::vector<int8_t> data_raw(src_data.size());
+                std::copy_n(src_data.data(), src_data.size(), data_raw.data());
+                return set_data_raw(element_offset, data_raw.data(), element_count);
+            }
+            case DataType::INT16: {
+                auto src_data = data->scalars().int_data().data();
+                std::vector<int16_t> data_raw(src_data.size());
+                std::copy_n(src_data.data(), src_data.size(), data_raw.data());
```
Comment (cydrain) [inline internal/core/src/segcore/ConcurrentVector.h:118]: why INT8/INT16 need copy_n ?

File: internal/core/src/segcore/ConcurrentVector.h [hunk 20]
```diff
@@ -84,6 +90,113 @@ class VectorBase {
     virtual void
     set_data_raw(ssize_t element_offset, const void* source, ssize_t element_count) = 0;
 
+    void
+    set_data_raw(ssize_t element_offset, ssize_t element_count, const DataArray* data, const FieldMeta& field_meta) {
```
Comment (cydrain) [inline internal/core/src/segcore/ConcurrentVector.h:94]: this API should be moved to ConcurrentVector.cpp

File: internal/core/src/segcore/ConcurrentVector.h [hunk 21]
```diff
@@ -144,16 +266,26 @@ class ConcurrentVectorImpl : public VectorBase {
     }
 
     void
-    set_data_raw(ssize_t element_offset, const void* source, ssize_t element_count) override {
-        set_data(element_offset, static_cast<const Type*>(source), element_count);
+    fill_chunk_data(const void* source, ssize_t element_count) override {
```
Comment (cydrain) [inline internal/core/src/segcore/ConcurrentVector.h:269]: this API should be moved to ConcurrentVector.cpp

File: internal/core/src/segcore/ConcurrentVector.h [hunk 22]
```diff
@@ -84,6 +90,113 @@ class VectorBase {
...
+                PanicInfo("unsupported");
+            }
+        }
+    }
+
+    virtual void
+    fill_chunk_data(const void* source, ssize_t element_count) = 0;
+
+    void
+    fill_chunk_data(ssize_t element_count, const DataArray* data, const FieldMeta& field_meta) {
```
Comment (cydrain) [inline internal/core/src/segcore/ConcurrentVector.h:149]: this API is too long, should be moved to ConcurrentVector.cpp

File: internal/core/src/segcore/ConcurrentVector.h [hunk 23]
```diff
@@ -144,16 +266,26 @@ class ConcurrentVectorImpl : public VectorBase {
...
+            return;
+        }
+        AssertInfo(chunks_.size() == 0, "no empty concurrent vector");
+        chunks_.emplace_to_at_least(1, Dim * element_count);
+        set_data(0, static_cast<const Type*>(source), element_count);
     }
 
     void
-    set_data(ssize_t element_offset, const Type* source, ssize_t element_count) {
+    set_data_raw(ssize_t element_offset, const void* source, ssize_t element_count) override {
```
Comment (cydrain) [inline internal/core/src/segcore/ConcurrentVector.h:279]: move all implementation to .cpp

File: internal/core/src/segcore/FieldIndexing.h [hunk 24]
```diff
@@ -118,22 +118,22 @@ class IndexingRecord {
     void
     Initialize() {
         int offset_id = 0;
-        for (const FieldMeta& field : schema_) {
-            auto offset = FieldOffset(offset_id);
+        for (auto& field : schema_.get_fields()) {
+            auto& field_meta = field.second;
```
Comment (cydrain) [inline internal/core/src/segcore/FieldIndexing.h:122]: also add: 
auto& field_id = field.first;

File: internal/core/src/segcore/SegmentGrowingImpl.cpp [hunk 25]
```diff
@@ -105,121 +106,84 @@ SegmentGrowingImpl::mask_with_delete(BitsetType& bitset, int64_t ins_barrier, Ti
...
-    auto raw_data = reinterpret_cast<const char*>(entities_raw.raw_data);
-    auto len_per_row = entities_raw.sizeof_per_row;
-    std::vector<std::tuple<Timestamp, idx_t, int64_t>> ordering;
-    ordering.resize(size);
-    // #pragma omp parallel for
-    for (int i = 0; i < size; ++i) {
-        ordering[i] = std::make_tuple(timestamps_raw[i], uids_raw[i], i);
-    }
-    std::sort(ordering.begin(), ordering.end());
+    // sort timestamps when flow graph receive insert msgs
```
Comment (cydrain) [inline internal/core/src/segcore/SegmentGrowingImpl.cpp:128]: comments need update

File: internal/core/src/common/Types.h [hunk 26]
```diff
@@ -36,19 +39,28 @@ constexpr auto MAX_TIMESTAMP = std::numeric_limits<Timestamp>::max();
...
 using engine::DataType;
 using engine::idx_t;
+constexpr auto MAX_ROW_COUNT = std::numeric_limits<engine::idx_t>::max();
 
 using ScalarArray = proto::schema::ScalarField;
 using DataArray = proto::schema::FieldData;
 using VectorArray = proto::schema::VectorField;
 using IdArray = proto::schema::IDs;
 using MetricType = faiss::MetricType;
+using InsertData = proto::segcore::InsertRecord;
```
Comment (cydrain) [inline internal/core/src/common/Types.h:49]: why not use "InsertRecord" ?

File: internal/core/src/segcore/SegmentGrowingImpl.cpp [hunk 27]
```diff
@@ -105,121 +106,84 @@ SegmentGrowingImpl::mask_with_delete(BitsetType& bitset, int64_t ins_barrier, Ti
...
-                              const std::vector<aligned_vector<uint8_t>>& columns_data) {
     // step 4: fill into Segment.ConcurrentVector
-    record_.timestamps_.set_data(reserved_begin, timestamps, size);
-    record_.uids_.set_data(reserved_begin, row_ids, size);
-    for (int fid = 0; fid < schema_->size(); ++fid) {
-        auto field_offset = FieldOffset(fid);
-        record_.get_field_data_base(field_offset)->set_data_raw(reserved_begin, columns_data[fid].data(), size);
+    insert_record_.timestamps_.set_data_raw(reserved_offset, timestamps_raw, size);
+    insert_record_.row_ids_.set_data_raw(reserved_offset, row_ids, size);
+    for (auto [field_id, field_meta] : schema_->get_fields()) {
```
Comment (cydrain) [inline internal/core/src/segcore/SegmentGrowingImpl.cpp:137]: you can iterate "insert_data->fields_data()" instead of "schema_->get_fields()", 
in this way we need not generate "field_it_to_offset"

File: internal/core/src/segcore/SegmentGrowingImpl.cpp [hunk 28]
```diff
@@ -105,121 +106,84 @@ SegmentGrowingImpl::mask_with_delete(BitsetType& bitset, int64_t ins_barrier, Ti
...
-    } else {
-        auto offset = schema_->get_primary_key_offset().value_or(FieldOffset(-1));
-        AssertInfo(offset.get() != -1, "Primary key offset is -1");
-        auto& row = columns_data[offset.get()];
-        auto row_ptr = reinterpret_cast<const int64_t*>(row.data());
-        for (int i = 0; i < size; ++i) {
-            uid2offset_.insert(std::make_pair(row_ptr[i], reserved_begin + i));
-        }
+    // step 5: set pks to offset
+    auto field_id = schema_->get_primary_field_id().value_or(FieldId(-1));
```
Comment (cydrain) [inline internal/core/src/segcore/SegmentGrowingImpl.cpp:145]: should use "INVALID_ID" instead of "-1"

File: internal/proxy/plan_parser.go (hunk 14 above)
Comment (DragonDriver) [reply internal/proxy/plan_parser.go:240]: We can simply support the basic expression in this patch and then extend the grammar rules of expression later.
For now, `startsWith` and `endsWith` are the default supported version in our parser library.

File: internal/core/src/index/StringIndexSort.h (hunk 16 above)
Comment (DragonDriver) [reply internal/core/src/index/StringIndexSort.h:34]: For historical reason, the original interface definition of "Index" was what you can see now. If we change it here, we'll change a lot of places that call it, which in fact is not very related to this topic.

Comment (DragonDriver) [reply internal/core/src/index/StringIndexSort.h:34]: Maybe we can change this in later pr.

File: internal/core/src/common/FieldMeta.h (hunk 1 above)
Comment (xige-16) [reply internal/core/src/common/FieldMeta.h:143]: The type of string_info_ is std::optional<StringInfo> , and this way(string_info_(max_length_per_row)) not work

File: internal/core/src/common/Schema.h (hunk 2 above)
Comment (xige-16) [reply internal/core/src/common/Schema.h:152]: fix

File: internal/core/src/common/Schema.cpp (hunk 3 above)
Comment (xige-16) [reply internal/core/src/common/Schema.cpp:74]: fix

File: internal/core/src/common/Schema.h (hunk 4 above)
Comment (xige-16) [reply internal/core/src/common/Schema.h:117]: This pr is relatively large, and it is too troublesome to split out another pr now.

File: internal/core/src/common/Types.cpp (hunk 5 above)
Comment (xige-16) [reply internal/core/src/common/Types.cpp:60]: DataType::STRING is not supported now.

File: internal/core/src/common/Types.cpp (hunk 6 above)
Comment (xige-16) [reply internal/core/src/common/Types.cpp:66]: fix

File: internal/core/src/common/Utils.h (hunk 8 above)
Comment (xige-16) [reply internal/core/src/common/Utils.h:22]: fix

File: internal/core/src/common/Utils.h (hunk 9 above)
Comment (xige-16) [reply internal/core/src/common/Utils.h:35]: fix

File: internal/core/src/segcore/FieldIndexing.cpp (hunk 10 above)
Comment (xige-16) [reply internal/core/src/segcore/FieldIndexing.cpp:152]: fix

File: internal/core/src/utils/Types.h (hunk 11 above)
Comment (xige-16) [reply internal/core/src/utils/Types.h:44]: fix

File: internal/core/thirdparty/knowhere/CMakeLists.txt (hunk 12 above)
Comment (xige-16) [reply internal/core/thirdparty/knowhere/CMakeLists.txt:30]: fix

File: internal/core/unittest/bench/bench_indexbuilder.cpp (hunk 13 above)
Comment (xige-16) [reply internal/core/unittest/bench/bench_indexbuilder.cpp:62]: fix

File: internal/core/src/common/FieldMeta.h (hunk 15 above)
Comment (xige-16) [reply internal/core/src/common/FieldMeta.h:159]: datatype_is_string() is a util function, not a member function of fieldMeta

File: internal/core/src/segcore/ConcurrentVector.h (hunk 19 above)
Comment (xige-16) [reply internal/core/src/segcore/ConcurrentVector.h:118]: protobuf does not support int8/int16 and does type promotion automatically

File: internal/core/src/segcore/ConcurrentVector.h (hunk 20 above)
Comment (xige-16) [reply internal/core/src/segcore/ConcurrentVector.h:94]: fix

File: internal/core/src/segcore/FieldIndexing.h (hunk 24 above)
Comment (xige-16) [reply internal/core/src/segcore/FieldIndexing.h:122]: fix

File: internal/core/src/segcore/SegmentGrowingImpl.cpp (hunk 25 above)
Comment (xige-16) [reply internal/core/src/segcore/SegmentGrowingImpl.cpp:128]: has updated

File: internal/core/src/common/Types.h (hunk 26 above)
Comment (xige-16) [reply internal/core/src/common/Types.h:49]: There is already a class nema InsertRecord

File: internal/core/src/segcore/SegmentGrowingImpl.cpp (hunk 27 above)
Comment (xige-16) [reply internal/core/src/segcore/SegmentGrowingImpl.cpp:137]: field_id_to_offset is also used to check if there is a duplicate field_id in the inserted data

File: internal/core/src/segcore/SegmentGrowingImpl.cpp (hunk 28 above)
Comment (xige-16) [reply internal/core/src/segcore/SegmentGrowingImpl.cpp:145]: fix

File: internal/core/src/segcore/ConcurrentVector.h (hunk 22 above)
Comment (xige-16) [reply internal/core/src/segcore/ConcurrentVector.h:149]: fix

File: internal/core/src/segcore/ConcurrentVector.h (hunk 23 above)
Comment (xige-16) [reply internal/core/src/segcore/ConcurrentVector.h:279]: This is the member function of the template class

File: internal/core/src/segcore/ConcurrentVector.h (hunk 21 above)
Comment (xige-16) [reply internal/core/src/segcore/ConcurrentVector.h:269]: This is the member function of the template class

File: internal/core/src/index/StringIndexSort.h (hunk 18 above)
Comment (DragonDriver) [reply internal/core/src/index/StringIndexSort.h:54]: Yes, we'll change this all-index-related codes in later pr.

File: internal/core/src/index/StringIndexSort.h (hunk 17 above)
Comment (DragonDriver) [reply internal/core/src/index/StringIndexSort.h:44]: Yes, we'll change this all-index-related codes in later pr.

File: internal/core/unittest/test_index_wrapper.cpp [hunk 29]
```diff
@@ -76,14 +76,14 @@ class IndexWrapperTest : public ::testing::TestWithParam<Param> {
 
         auto dataset = GenDataset(NB, metric_type, is_binary);
         if (!is_binary) {
-            xb_data = dataset.get_col<float>(0);
+            xb_data = dataset.get_col<float>(milvus::FieldId(100));
```
Comment (cydrain) [inline internal/core/unittest/test_index_wrapper.cpp:79]: Use "START_USER_FIELDID" instead of "100", and check all these places

File: internal/util/typeutil/type.go [hunk 30]
```diff
@@ -49,3 +49,5 @@ const (
 )
 
 const InvalidUniqueID = UniqueID(-1)
+const InvalidIntPK = int64(-1)
```
Comment (xiaofan-luan) [inline internal/util/typeutil/type.go:52]: We did restrict that PK can not be nagative or ""?

File: internal/proxy/util.go [hunk 31]
```diff
@@ -22,16 +22,19 @@ import (
...
 	"github.com/milvus-io/milvus/internal/proto/schemapb"
-	"github.com/milvus-io/milvus/internal/util/funcutil"
 )
 
 // enableMultipleVectorFields indicates whether to enable multiple vector fields.
 const enableMultipleVectorFields = false
 
 // maximum length of variable-length strings
-const maxVarCharLength = int64(65535)
+const maxVarCharLengthKey = "max_length_per_row"
```
Comment (xiaofan-luan) [inline internal/proxy/util.go:36]: is it max_length_per_row or max_length_per_field?

Comment (xige-16) [reply internal/proxy/util.go:36]: > is it max_length_per_row or max_length_per_field?

Will max_length_per_field be misinterpreted as the maximum number of rows per field?

File: internal/core/src/segcore/Utils.cpp [hunk 32]
```diff
@@ -0,0 +1,258 @@
...
+
+// TODO remove merge dataArray, instead fill target entity when get data slice
+std::unique_ptr<DataArray>
+MergeDataArray(std::vector<std::pair<milvus::SearchResult*, int64_t>>& result_offsets, const FieldMeta& field_meta) {
+    auto data_type = field_meta.get_data_type();
+    auto data_array = std::make_unique<DataArray>();
+    data_array->set_field_id(field_meta.get_id().get());
+    data_array->set_type(milvus::proto::schema::DataType(field_meta.get_data_type()));
+
+    for (auto& result_pair : result_offsets) {
```
Comment (xiaofan-luan) [inline internal/core/src/segcore/Utils.cpp:186]: should not add 1 element each time.
Reserve result_offsets.size to avoid copy in add

File: internal/core/src/segcore/segment_c.cpp [hunk 33]
```diff
@@ -133,54 +134,16 @@ Insert(CSegmentInterface c_segment,
        int64_t size,
        const int64_t* row_ids,
        const uint64_t* timestamps,
-       void* raw_data,
-       int sizeof_per_row,
-       int64_t count) {
+       const char* data_info) {
```
Comment (xiaofan-luan) [inline internal/core/src/segcore/segment_c.cpp:137]: is there any performance overhead to change to protobuf format here?

File: internal/core/src/segcore/Utils.cpp [hunk 34]
```diff
@@ -0,0 +1,258 @@
...
+            break;
+        }
+        default: {
+            PanicInfo("unsupported");
+        }
+    }
+}
+
+void
+ParsePksFromIDs(std::vector<PkType>& pks, DataType data_type, const IdArray& data) {
```
[the remark this answers is not in the document]
Comment (xiaofan-luan) [reply internal/core/src/segcore/Utils.cpp:36]: what if iterate through IdArray and convert one by one

[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/segcore/Utils.cpp:36]: how to convert one by one from IdArray without copy?

File: internal/core/src/segcore/SegmentGrowingImpl.cpp [hunk 35]
```diff
@@ -364,79 +334,23 @@ SegmentGrowingImpl::bulk_subscript(SystemFieldType system_type,
...
-            column.insert(column.end(), beg, beg + element_sizeof);
-        }
-        columns_data.emplace_back(std::move(column));
-    }
-    do_insert(reserved_offset, size, row_ids.data(), timestamps.data(), columns_data);
-}
-
 std::vector<SegOffset>
 SegmentGrowingImpl::search_ids(const BitsetType& bitset, Timestamp timestamp) const {
     std::vector<SegOffset> res_offsets;
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/segcore/SegmentGrowingImpl.cpp:346]: > res_offsets should be preallocated to bitset.cardinality to avoid copy

This size cannot be known in advance.

File: internal/core/src/segcore/load_index_c.cpp [hunk 36]
```diff
@@ -107,3 +110,39 @@ AppendIndex(CLoadIndexInfo c_load_index_info, CBinarySet c_binary_set) {
...
+        auto load_index_info = (LoadIndexInfo*)c_load_index_info;
+        auto field_type = load_index_info->field_type;
+        auto binary_set = (knowhere::BinarySet*)c_binary_set;
+        auto& index_params = load_index_info->index_params;
+        // TODO(xige-16) check index_type
+        // bool find_index_type = index_params.count("index_type") > 0 ? true : false;
+        // AssertInfo(find_index_type == true, "Can't find index type in index_params");
+        // load_index_info->index =
+        //     milvus::scalar::IndexFactory::GetInstance().CreateIndex(field_type, index_params["index_type"]);
+        load_index_info->index = milvus::scalar::IndexFactory::GetInstance().CreateIndex(field_type, "");
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/segcore/load_index_c.cpp:126]: When building a scalar index, there is no requirement to specify index_type

File: internal/core/src/index/StringIndexSort.h [hunk 37]
```diff
@@ -0,0 +1,72 @@
...
+    PrefixMatch(std::string prefix) {
+        auto data = GetData();
+        TargetBitmapPtr bitset = std::make_unique<TargetBitmap>(data.size());
+        for (size_t i = 0; i < data.size(); i++) {
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/index/StringIndexSort.h:58]: > Since the data is already sorted in ScalarIndexSort, use a binary search should accelerate search spped

Maybe we can change this in later pr.

File: internal/core/src/segcore/ConcurrentVector.h [hunk 38]
```diff
@@ -66,6 +65,13 @@ class ThreadSafeVector {
...
+    void
+    clear() {
+        std::lock_guard lck(mutex_);
+        size_ = 0;
+        vec_.clear();
+    }
+
  private:
     std::atomic<int64_t> size_ = 0;
     std::deque<Type> vec_;
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/segcore/ConcurrentVector.h:77]: > Question, why deque here? random access would be really slow for a deque. vector or some data structure such as folly::concurrentmap make more sense to me

no random access usage

File: internal/core/src/segcore/load_index_c.cpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/core/src/segcore/load_index_c.cpp]: fixed

File: internal/core/src/segcore/Utils.cpp (hunk 32 above)
Comment (xige-16) [reply internal/core/src/segcore/Utils.cpp:186]: I'll change this in next pr.

File: internal/util/typeutil/type.go (hunk 30 above)
Comment (xige-16) [reply internal/util/typeutil/type.go:52]: fixed

File: internal/core/src/segcore/ConcurrentVector.h (hunk 38 above)
[the remark this answers is not in the document]
Comment (xiaofan-luan) [reply internal/core/src/segcore/ConcurrentVector.h:77]: operator[] is random access, and it is used when get chunk

File: internal/core/src/segcore/SegmentGrowingImpl.cpp (hunk 35 above)
[the remark this answers is not in the document]
Comment (xiaofan-luan) [reply internal/core/src/segcore/SegmentGrowingImpl.cpp:350]: bitset.count would help, most bitset has a counter about how many ones or zeros are there in a bit set
